Task 2

8Bts opened the pull request:
Hi @amikishiyev.

I tried to rebase on the new branch and I got too many merge conflicts. So I just branched off from **task** and after committing to it I compared the new branch to the previous one. Only the latest changes reflect now. Hope it will be easier to review.
Here are the major changes I made:

- Added flash on success and listed errors on failure
- Configured `sidekiq` adapter for job queue (for production env)
-  Wrapped API call in the `Api` module
- Added request specs for testing API calls with VCR
- Added controller specs

8Bts committed:
Add flash message and error for view

8Bts committed:
Change queue adapter to use sidekiq

8Bts committed:
Add request and controller specs

amikishiyev commented on spec/controllers/authors_controller_spec.rb:
```diff
@@ -0,0 +1,55 @@
+             }
+      end
+
+      it { should render_template(:new) }
```
Each of these 3 examples will run separate post call. Consider optimizing tests for all possible minimums.

amikishiyev commented on spec/jobs/request_job_spec.rb:
```diff
@@ -18,7 +18,7 @@
       ActiveJob::Base.queue_adapter = :test
```
Useless test because it will never break. Accept it as a rule, there is no need to test rails functionality in your application tests.

amikishiyev commented on app/jobs/request_job.rb:
```diff
@@ -16,24 +17,6 @@ class RequestJob < ApplicationJob
-
-    puts "status #{response.status}"
-    puts JSON.parse(response.body)
+    Api.send_data(params)
```
What was the reason for extracting this as a module?

8Bts replied:
Just for making it reusable and easy to test. I believe that wrapping API implementation is a good practice as it gives you plenty of reasonable benefits.

amikishiyev commented on spec/controllers/authors_controller_spec.rb:
```diff
@@ -0,0 +1,55 @@
+require 'rails_helper'
+
+RSpec.describe AuthorsController, type: :controller do
```
We still miss the test for making sure that it calls the `RequestJob.perform_later` if params are valid. Another useful test is to ensure that it doesn't call `RequestJob.perform_later` if params are invalid (just one combination of wrong input will be enough).

amikishiyev commented on spec/controllers/authors_controller_spec.rb:
```diff
@@ -0,0 +1,55 @@
+        post :create, params: nil
+        expect(response).to have_http_status(422)
+
+        post :create, params: {
```
I think there is no need to test every combination of wrong input here as you already covered all cases in Author model.

amikishiyev commented on spec/requests/api_request_spec.rb:
```diff
@@ -1,50 +1,31 @@
 require 'rails_helper'
+require_relative '../../app/api/api'
```
wouldn't it work without this include?

amikishiyev commented on app/views/authors/new.html.erb:
```diff
@@ -1,12 +1,25 @@
-  <%=label_tag 'first_name' %>
-  <%=text_field_tag 'first_name' %>
+<div>
+  <% if flash[:success] %>
```
Wouldn't rails handle all this newly added staff if you use `form_for` helper instead of `form_tag`? I realize @author is not an instance of ApplicationRecord but it seems including some additional modules in Author class would save us from this manual processing.

amikishiyev commented on spec/rails_helper.rb:
```diff
@@ -38,7 +38,7 @@
   # instead of true.
   config.use_transactional_fixtures = true
-
+  config.render_views
```
including this globally is a bad idea as rendering views is expensive operation and in large projects unnecessarily complicate the data preparation.

amikishiyev commented:
Is it clear that you was actually asked to work on two separate tasks? The first task in this rails project. 

For the second task the deliverables are: a custom ruby class and tests suite for it.

8Bts committed:
Change form_tag helper to form_for

8Bts committed:
Refactor `authors_controller_spec.rb`

8Bts committed:
Refactor `api_request_spec.rb` to use webmock without VCR

8Bts committed:
Refactor job specs

8Bts commented:
Hi @amikishiyev. 
Sorry for the late reply. I was really busy and couldn't work for 3 days.  
Check out the new changes I made:

1.  `form_tag` replaced with `form_for` helper.
2.  Added tests for the `perform_later` call in the controller.
3.  Replaced VCR with webmock where real HTTP requests performed.
4.  Included only one test in job specs for API calls.
